File: packages/agent-runtime/src/adapters/ollama-provider.ts
/**
 * @ainative/agent-runtime — OllamaProvider
 * Built by AINative Dev Team
 * Refs #248
 *
 * Wraps the Ollama /api/chat endpoint to match the LLMProvider interface.
 */

import { v4 as uuidv4 } from 'uuid';
import type {
  LLMProvider,
  LLMMessage,
  LLMToolDefinition,
  LLMChatOptions,
  LLMResponse,
  ToolCall,
} from '../types';

// ─── Config ───────────────────────────────────────────────────────────────────

export interface OllamaProviderConfig {
  baseUrl?: string;
  model?: string;
}

// ─── Ollama wire types ────────────────────────────────────────────────────────

interface OllamaToolFunction {
  name: string;
  arguments: Record<string, unknown>;
}

interface OllamaToolCall {
  function: OllamaToolFunction;
}

interface OllamaMessage {
  role: string;
  content: string;
  tool_calls?: OllamaToolCall[];
}

interface OllamaChatResponse {
  model: string;
  message: OllamaMessage;
  done: boolean;
}

interface OllamaToolDefinition {
  type: 'function';
  function: {
    name: string;
    description: string;
    parameters?: Record<string, unknown>;
  };
}

// ─── OllamaProvider ──────────────────────────────────────────────────────────

export class OllamaProvider implements LLMProvider {
  readonly baseUrl: string;
  readonly model: string;
  /** Human-readable name for use in model selection */
  readonly name: string;

  constructor(config: OllamaProviderConfig) {
    this.baseUrl = config.baseUrl ?? 'http://localhost:11434';
    this.model = config.model ?? 'llama3.2';
    this.name = `ollama:${this.model}`;
  }

  // ─── chat() ───────────────────────────────────────────────────────────────

  async chat(messages: LLMMessage[], options?: LLMChatOptions): Promise<LLMResponse> {
    const body: Record<string, unknown> = {
      model: this.model,
      messages,
      stream: false,
    };

    if (options?.temperature !== undefined) {
      body.options = { temperature: options.temperature };
    }

    const response = await fetch(`${this.baseUrl}/api/chat`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      const text = await response.text();
      throw new Error(`Ollama error ${response.status}: ${text}`);
    }

    const data = (await response.json()) as OllamaChatResponse;
    return {
      content: data.message.content ?? '',
      toolCalls: [],
    };
  }

  // ─── chatWithTools() ──────────────────────────────────────────────────────

  async chatWithTools(
    messages: LLMMessage[],
    tools: LLMToolDefinition[],
    options?: LLMChatOptions,
  ): Promise<LLMResponse> {
    const ollamaTools: OllamaToolDefinition[] = tools.map((t) => ({
      type: 'function',
      function: {
        name: t.name,
        description: t.description,
        parameters: t.parameters,
      },
    }));

    const body: Record<string, unknown> = {
      model: this.model,
      messages,
      tools: ollamaTools,
      stream: false,
    };

    if (options?.temperature !== undefined) {
      body.options = { temperature: options.temperature };
    }

    const response = await fetch(`${this.baseUrl}/api/chat`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      const text = await response.text();
      throw new Error(`Ollama error ${response.status}: ${text}`);
    }

    const data = (await response.json()) as OllamaChatResponse;

    const toolCalls: ToolCall[] = (data.message.tool_calls ?? []).map((tc) => ({
      id: uuidv4(),
      name: tc.function.name,
      args: tc.function.arguments,
    }));

    return {
      content: data.message.content ?? '',
      toolCalls,
    };
  }
}

File: packages/agent-runtime-py/ainative_agent_runtime/adapters/ollama_provider.py
"""
ainative-agent-runtime — OllamaProvider (Python)
Built by AINative Dev Team
Refs #248

Wraps the Ollama /api/chat endpoint to match the LLMProvider protocol.
"""

from __future__ import annotations

import json
import uuid
from typing import Any, Dict, List, Optional

import aiohttp


class OllamaProvider:
    """
    LLM provider adapter for locally-running Ollama instances.

    Args:
        base_url: Base URL of the Ollama server (default: http://localhost:11434).
        model: Model name to use (default: llama3.2).
    """

    def __init__(
        self,
        base_url: str = "http://localhost:11434",
        model: str = "llama3.2",
    ) -> None:
        self.base_url = base_url
        self.model = model
        self.name = f"ollama:{model}"

    # ─── Internal HTTP helper ─────────────────────────────────────────────────

    async def _post_chat(self, body: Dict[str, Any]) -> Dict[str, Any]:
        async with aiohttp.ClientSession() as session:
            async with session.post(
                f"{self.base_url}/api/chat",
                data=json.dumps(body),
                headers={"Content-Type": "application/json"},
            ) as response:
                if response.status != 200:
                    text = await response.text()
                    raise RuntimeError(f"Ollama error {response.status}: {text}")
                return await response.json()

    # ─── chat() ───────────────────────────────────────────────────────────────

    async def chat(
        self,
        messages: List[Dict[str, Any]],
        options: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        body: Dict[str, Any] = {
            "model": self.model,
            "messages": messages,
            "stream": False,
        }
        if options and options.get("temperature") is not None:
            body["options"] = {"temperature": options["temperature"]}

        data = await self._post_chat(body)
        return {
            "content": data["message"].get("content", ""),
            "tool_calls": [],
        }

    # ─── chat_with_tools() ────────────────────────────────────────────────────

    async def chat_with_tools(
        self,
        messages: List[Dict[str, Any]],
        tools: List[Dict[str, Any]],
        options: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        ollama_tools = [
            {
                "type": "function",
                "function": {
                    "name": t["name"],
                    "description": t["description"],
                    "parameters": t.get("parameters"),
                },
            }
            for t in tools
        ]

        body: Dict[str, Any] = {
            "model": self.model,
            "messages": messages,
            "tools": ollama_tools,
            "stream": False,
        }
        if options and options.get("temperature") is not None:
            body["options"] = {"temperature": options["temperature"]}

        data = await self._post_chat(body)
        raw_calls = data["message"].get("tool_calls") or []

        tool_calls = [
            {
                "id": str(uuid.uuid4()),
                "name": tc["function"]["name"],
                "args": tc["function"].get("arguments", {}),
            }
            for tc in raw_calls
        ]

        return {
            "content": data["message"].get("content", ""),
            "tool_calls": tool_calls,
        }
